docs/design/mesos-style.md

```diff
@@ -0,0 +1,162 @@
+a Job. Borg has corresponding state machine logic for managing Jobs and their instances, and a scheduler that's responsible
+for assigning the instances to machines.
+
+The flow of a request in Borg is:
+
+1. Client submits a collection object to the Borgmaster API endpoint
+1. Admission control, quota, applying defaults, etc. run on the collection
+1. If the collection is admitted, it is persisted, and the collection state machine creates the underlying instances
+1. The scheduler assigns a hostname to the instance, and tells the Borglet to start the instance's container(s)
+1. Borglet starts the container(s)
+1. The instance state machine manages the instances and the collection state machine manages the collection during their lifetimes
+
+Out-of-the-box Kubernetes has *workload-specific* abstractions (ReplicaSet, Job, DaemonSet, etc.) and corresponding controllers,
+and in the future may have [workload-specific schedulers](../../docs/proposals/multiple-schedulers.md),
+e.g. different schedulers for long-running services vs. short-running batch. But these abstractions, controllers, and
+schedulers are not *application-specific*.
+
+The usual request flow in Kubernetes is very similar, namely
+
+1. Client submits a collection object (e.g. ReplicaSet, Job, ...) to the API server
+1. Admission control, quota, applying defaults, etc. run on the collection
+1. If the collection is admitted, it is persisted, and the corresponding collection controller creates the underlying pods
+1. Admission control, quota, applying defaults, etc. runs on each pod; if there are multiple schedulers, one of the admission
+controllers will write the scheduler name as an annotation based on a policy
+1. If a pod is admitted, it is persisted
+1. The appropriate scheduler assigns a nodeName to the instance, which triggers the Kubelet to start the pod's container(s)
+1. Kubelet starts the container(s)
+1. The controller corresponding to the collection manages the pod and the collection during their lifetime
+
+Because Borg-style abstractions are not application-specific, it is common to end up with "two-level scheduling" when
+using Borg. For example, the MapReduce master might create a Borg Job with a certain number of instances with a particular resource shape,
+and then schedule the individual units of map and reduce work onto those instances itself.
+
+## What is a Mesos-style architecture?
+
+Mesos is fundamentally designed to support multiple application-specific "frameworks." A framework is
+composed of a "framework scheduler" and a "framework executor." We will abbreviate "framework scheduler"
+as "framework" since "scheduler" means something very different in Kubernetes (something that just
+assigns pods to nodes).
+
+Unlike Borg and Kubernetes, where there is a single logical endpoint that receives all API requests (the Borgmaster and API server,
```

Comment:
agree that a more nuanced description is needed here. controllers in k8s-land can provide endpoints too. kubelets provide endpoints. it's a bit disingenuous to write that mesos API's are fragmented because frameworks can provide API endpoints.

mesos itself provides an HTTP-based API (https://github.com/apache/mesos/blob/master/docs/scheduler-http-api.md) that "clients" (frameworks) use to schedule Tasks (the execution primitive in mesos). mesos also provides hooks for authorization and authentication that have begun to (and are planned to completely) integrate with the exposed HTTP API. frameworks that want fine-grained control over Task execution can implement custom executors, for which mesos also provides an HTTP API (https://github.com/apache/mesos/blob/master/docs/executor-http-api.md).